File: c++/数据结构/字典树/01trie.cpp
struct Trie
{
	#define type int
	static const int LOG=30;
	static const int K=LOG+2;
	static const type inf=INF;
	int rt,root[MAX],tot,ch[MAX*K][2],cnt[MAX*K];
	Trie &operator[](const int _rt){this->rt=_rt;return *this;}
	int newnode()
	{
		tot++;
		memset(ch[tot],0,sizeof ch[tot]);
		cnt[tot]=0;
		return tot;
	}
	void init(int n=1)
	{
		ch[0][0]=ch[0][1]=cnt[0]=0;
		tot=1;
		for(int i=0;i<=n;i++) root[i]=0;
		rt=1;
	}
	void insert(type x)
	{
		int id,t,i;
		if(!root[rt]) root[rt]=newnode();
		id=root[rt];
		for(i=LOG;~i;i--)
		{
			cnt[id]++;
			t=(x>>i)&1;
			if(!ch[id][t]) ch[id][t]=newnode();
			id=ch[id][t];
		}
		cnt[id]++;
	}
	type ask_max(type x)
	{
		int id,i;
		type res,t;
		if(!root[rt]) return -inf;
		id=root[rt];
		res=0;
		for(i=LOG;~i;i--)
		{
			t=(x>>i)&1;
			if(ch[id][t^1]&&cnt[ch[id][t^1]]) t^=1;
			res|=(t<<i);
			id=ch[id][t];
		}
		return res^x;
	}
	type ask_min(type x)
	{
		int id,i;
		type res,t;
		if(!root[rt]) return inf;
		id=root[rt];
		res=0;
		for(i=LOG;~i;i--)
		{
			t=(x>>i)&1;
			if(!ch[id][t]||!cnt[ch[id][t]]) t^=1;
			res|=(t<<i);
			id=ch[id][t];
		}
		return res^x;
	}
	#undef type
}tr;

File: kotlin/01Trie.kt
class Trie
{
	var LOG=30
	var K=LOG+2
	var rt=0
	var root=IntArray(MAX){0}
	var tot=0
	var ch=Array(MAX*K){IntArray(2){0}}
	var v=LongArray(MAX*K){0L}
	fun newnode():Int
	{
		tot++
		ch[tot][0]=0
		ch[tot][1]=0
		v[tot]=0
		return tot
	}
	fun init()
	{
		var n=1 //
		ch[0][0]=0
		ch[0][1]=0
		v[0]=0L
		tot=1
		for(i in 0..n) root[i]=0
		rt=1
	}
	fun insert(x:Long,qv:Long)
	{
		if(root[rt]==0) root[rt]=newnode()
		var id=root[rt]
		for(i in LOG downTo 0)
		{
			v[id]+=qv
			var t=((x shr i) and 1).toInt()
			if(ch[id][t]==0) ch[id][t]=newnode()
			id=ch[id][t]
		}
		v[id]+=qv
	}
	fun count_y(x:Long,limt:Long):Long // count y: x xor y <= limt
	{
		if(root[rt]==0) return 0
		var id=root[rt]
		var res=0L
		for(i in LOG downTo 0)
		{
			var t=((x shr i) and 1).toInt()
			if((limt shr i) and 1 > 0)
			{
				res+=v[ch[id][t]]
				id=ch[id][t xor 1]
			}
			else id=ch[id][t]
		}
		res+=v[id]
		return res
	}
}
